lib/elixir/lib/code/formatter.ex

```diff
@@ -656,6 +656,14 @@ defmodule Code.Formatter do
   defp block_next_line(:@), do: @empty
   defp block_next_line(_), do: break("")
 
+  defp maybe_add_newlines({:def, [{:@, _, _} = _next | _]}, _meta) do
```

Comment:
We should not look at `def`, because we can have `def`, `defp`, `defmacro`, `defmacrop`... and projects can define `defn`, `defnp`, etc. We do have some code today that skips adding new lines if they are module attributes but perhaps the code should only avoid adding lines below, not lines up.

Reply:
Ah, that makes a lot of sense.

I'll try to play around with this in the next couple of days, but I won't be upset if you solve this in 2 minutes :)

No hard feelings :)

Reply:
@josevalim, seems closer now, but likely a bit far from the optimal solution.

I kinda wanted to offer a devil's advocate here, to the issue we put together :joy:. After reading and working on the code a little bit, I came to appreciate the `user's choice`. I didn't think of this when seeing the issue for the first time mostly because I had not seen/worked with/knew how the formatter works.

I like how the formatter respect's the user's choice __first__. The crux to this devil's advocate argument is that the original code "wanted" to save space by doing the shorthard def.

If space was not an issue, the code would have used a multiline def, which the formatter already handles:
```elixir
# pre-format
defmodule Example do
  @impl FooA
  def a() do
    :something
  end
  @impl FooB
  def b() do
    :something_else
  end
end

# post-format
defmodule Example do
  @impl FooA
  def a() do
    :something
  end

  @impl FooB
  def b() do
    :something_else
  end
end
```

Even after writing the above I'm more convinced that this is:
- not applicable
- likely a doc adjustment to document the `user's choice` rather than a behavior change

WDYT?

Reply:
I agree, looking at all the [formatting changes in this repo](https://github.com/elixir-lang/elixir/actions/runs/11129318552/job/30959556934?pr=13867) highlighted by the CI, this seems to have a much bigger impact than just the original example, and the user's choice perspective is a good point.